### ISSUE TITLE
[Bug]: Client will be broken soon

### ISSUE THREAD

**Issue by Palmke** [bropat/eufy-security-client#259]
### Client version

n.a.

### Node version

n.a.

### Operating System type

Other

### Operating system version

n.a.

### Describe the bug

Currently tracking the work of https://mobile.twitter.com/Paul_Reviews, he states requests from the web portal are now encrypted.
I noticed on mysecurity.eufylife.com that indeed the responses of `get_hub_list`, `get_devs_list` and `get_webrtc_configuration` are now encrypted.
I assume this can be easily solved because the decryption takes place on the web page itself. 

Just a heads up as everyone (re-)starting the client will have troubles.

### To reproduce

Logon to eufy webportal.

**Comment by Palmke** [comment]
@martijnpoppen You're a front-end developer, right?
Can't you look in the code of the webportal to see if you can find some decryption logic?

**Comment by martijnpoppen** [comment]
As far as I see this is a encryption made with crypto.

So you need a secret key to unlock this.
I did the same for multiple of my apps. 

Example: 
![image](https://user-images.githubusercontent.com/7694138/204296416-996938db-0d44-478a-bd07-82187ba9dd2f.png)
  
 After digging into the js of the mysecurity portal:
 
![image](https://user-images.githubusercontent.com/7694138/204296587-07243ebf-48df-4375-b74e-d1d15900e50d.png)

So my guess would be that we might find this key in the extracted APK as the same calls are made there right.

I'll check the portal a bit more :)

**Comment by martijnpoppen** [comment]
Ok got the key.
It's in the login response.

Will try to decrypt the response.

**Comment by Palmke** [comment]
lol this is so stupid.

Great work tho :-)

**Comment by bropat** [comment]
FYI: This client uses the Cloud API of the Android app and not the Web Portal (different APIs).

**Comment by Palmke** [comment]
That's still unencrypted? Regardless good to know they're again hammering down the API.

**Comment by bropat** [comment]
> That's still unencrypted?

Today a new version of the Android app was released (`v4.5.0_1510`). I will have a look at it...

**Comment by martijnpoppen** [comment]
Did make some progress on the decryption part. Will post a snippet once I have that working.

Nice good to know @bropat  :) 
Only the login is the same right ? the v2 login?

**Comment by bropat** [comment]
In the new version it looks like nothing has changed ;)

> Only the login is the same right ? the v2 login?

Honestly, I don't know, as I have never looked at the web portal :D
What I can say is that the client already uses the latest APIs for authentication and that the latest Android app version still uses the same calls.

e.g.:

POST /v2/passport/login_sec HTTP/2
```json
{
   "ab":"US",
   "client_secret_info":{
      "public_key":"04f10af16b3620bab8e09bd2840dca2c83a778cf21ed19034815a53001d7f36d6c5b685138a50ff8d3aa28b1edd811eaec9b4490d6636f80d1f6bd09f4fb68fffb"
   },
   "email":"<redacted>",
   "enc":0,
   "password":"<redacted>",
   "time_zone":3600000,
   "transaction":"624087956230"
}
```

Response:
```json
{
   "code":0,
   "msg":"Succeed.",
   "data":{
      "user_id":"<redacted>",
      "email":"<redacted>",
      "nick_name":"Abc***",
      "auth_token":"<redacted>",
      "token_expires_at":1672245838,
      "avatar":"",
      "invitation_code":"<redacted>",
      "inviter_code":"",
      "verify_code_url":"",
      "mac_addr":"",
      "domain":"security-app.eufylife.com",
      "ab_code":"US",
      "geo_key":"<redacted>",
      "privilege":0,
      "phone":"",
      "phone_code":"",
      "server_secret_info":{
         "public_key":"04a4baa7d15a42802e192fb679d08d49aeb6edce438a4062b060ea3f9dfd63521edd7940a215aec6dbdc75b5361e1647ac6577919f98a0690ff1b6225534f66e39"
      },
      "params":null,
      "trust_list":[
         
      ]
   }
}
```

**Comment by martijnpoppen** [comment]
Nice!
I won't expect that they'll encrypt the APP responses too. I think it was mainly due to this topic that they blocked encrypted it for the web version.

topic: https://community.security.eufy.com/t/privacy-quick-video-to-show-eufy-are-lying-about-devices-being-private-locally-stored-theyre-not/3527511

**Comment by bropat** [comment]
I agree with you.

**Comment by Palmke** [comment]
Cool, then that's settled for now. Apologies for stirring the kettle, though I do think we should keep an eye out for changes in the API.

**Comment by martijnpoppen** [comment]
So to get back on this.
Seems like Eufy decided to shut down the v1/web/equipment/start_stream and change that to a v2/web/equipment/start_stream which is encrypted.

Will check if I can fix it as the Homey integration does use this for streaming

**Comment by martijnpoppen** [comment]
@bropat do you know if theres also a start_stream call in the APK? Or is this webAPI only?

Looks like some AES encrypted response. But it's hard to find a decryption method for this.

**Comment by bropat** [comment]
@martijnpoppen 

> do you know if theres also a start_stream call in the APK? Or is this webAPI only?

Only web API.

**Comment by martijnpoppen** [comment]
Thats a shame. Did quite some digging into the webapi but cant find the encryption if the data field.

Looks like valid base64 but it decodees to unreadable chars

**Comment by rubenbroman** [comment]
@martijnpoppen I've been beating my head against the wall that is the minified js code of the eufy web site. There is a way to decode the data in there but I'm not smart/good enough with js to understand it. Hopefully someone else is?

**Comment by martijnpoppen** [comment]
@rubenbroman Yeah same. I went through all the minified stuff

Still looks like its something with AES or a Base64 encoded uint8array

**Comment by martijnpoppen** [comment]
@rubenbroman I got the decryption! Will try to make a PR with this decryption method today :)

**Comment by Palmke** [comment]
Out of curiosity, is it purely the REST data that is encrypted, or is the actual stream also encrypted?

**Comment by martijnpoppen** [comment]
@Palmke only the REST is encrypted. once you decrypt it you get the "normal" data.
Only struggling with the private key generation, but almost there :D

**Comment by bropat** [comment]
Eufy has now also brought some endpoints for the Android app up to "v2" and encrypted the response data:

![image](https://user-images.githubusercontent.com/2305255/207103147-8fb35b90-fd42-4a85-8426-d7ec044bffea.png)

![image](https://user-images.githubusercontent.com/2305255/207103227-148862fe-781f-42f5-b19f-af4d3259a783.png)

Eufy uses ECDH with the elyptic curve `P-256` for this purpose. The shared key is found in the login response:

```json
"server_secret_info":{"public_key":"0470ca96234263082120ab0646918187a9df3f0b70f1429e349b5d8a7bcde7cc4f5dd00fadbe1fd225ed0576fce145c785c8ed6b667485619bbcf57b2e5faa8405"},
```

Basically the same thing I have already implemented [here](https://github.com/bropat/eufy-security-client/blob/4d2684e56757b60c2c8afa1ff5c4df75016c6fac/src/http/api.ts#L261-L269), only that you have to decrypt `aes-256-cbc` instead of encrypting as linked above ;)

**Comment by martijnpoppen** [comment]
Oh thats interesting!
I already had a decryption for the webapp. There they store the keys in a base64 encoded array in de sessionstorage.

Unfortunately i only managed once to decrypt it. Will have a look!

Really dont like this change from Eufy :(

**Comment by bropat** [comment]
I have successfully carried out the decryption.
Will publish it in the `develop` branch in a short time.

**Comment by bropat** [comment]
@martijnpoppen 

Checkout [this commit](https://github.com/bropat/eufy-security-client/commit/f3090b2fa646363a3023b7e5c3209bd9ac2abc92).

I still have to "convert" the other endpoints. In the meantime, you can find the concept of the new communication in the commit. I think it won't be much different with the Web API ;)
Unfortunately I don't have more time today, but I'm sure it will help you ;)

Note on the side: I think the change from Eufy is not bad (more safety is always good). I think this change was already planned for a while, because the server public key has already been available for a few months for the login of the Eufy app and was only rolled out faster due to the media "problem" ;)

**Comment by martijnpoppen** [comment]
@bropat Awesome work! It also works for the web api and the stream is working again :)
I'll try to make a PR tonight 

note: Yes, I totally agree. Hopefully everything will keep working as we would like to 😄

**Comment by martijnpoppen** [comment]
Note: the stream response did change to this format:

```
{
  url: '',
  domain: 'mediaserver-usa3.eufylife.com',
  stream_name: 'REDACTED',
  time: '1670877344',
  token: 'REDACTED'
}
```

**Comment by Palmke** [comment]
I'm loving the reverse engineering going on here :D 
Great work, guys!
So this means the RTSP streams are finally encrypted?

**Comment by martijnpoppen** [comment]
Yeah without the decryption there's no way to get the stream url. But the stream url is still the same.

Yeah there are multiple servers. But all of them USA (mediaserver-3-usa mediaserver-4-usa pip-vir-7.eufylife pip-vir-8.eufylife)

**Comment by martijnpoppen** [comment]
unfortunately the stream doesn't work anymore :sad
Eufy set a Cors on the stream url.
